### DISCUSSION TITLE
H743 Nucleo boards

### DISCUSSION THREAD

**Discussion by dresco** [grblHAL/STM32F7xx#5] [General]
Hi Terje,

Looks like I'll have two of the Nucleo H743ZI boards from Monday (back orders fulfilled from two suppliers), can I send you one? 
If so, just let me know where to send it..

Cheers,
Jon

**Comment by terjeio** [comment]
I have been very busy lately so sorry for the late answer. I'll try to clear the issues backlog the coming days.

Thanks for the offer but IMO I believe it is best you keep the board as a backup? I have a large collection of different boards already and I am a bit hesitant to add to that since it might result in me providing support for the driver...

BTW is the low level lwIP driver provided for this chip any good? The F7xx driver is bad performance wise, IIRC an improved driver is to be released in Q2 or Q3. A good test is to time the transfer speed in check mode.

**Reply by dresco** [reply]
No worries, I appreciate the huge amount of effort you put into this project!

Yep - the latest ST package for the H7 included the rewritten Ethernet driver, so am using that. I hadn't got as far as any performance testing, but a very quick WS test just now shows roughly 10x improvement over default serial connection to the Nucleo for a large file (though that's still only ~1Mb/s). Do you have any benchmark files you use, else I'll just compare against Teensy performance?

To be honest, I was hoping that you would consider upstreaming the driver into the main repos - once I get it into a state that you are happy with. My motivation for the port was just to have something that gets close to Teensy performance, but with an improved development/debug experience.

**Reply by terjeio** [reply]
> Do you have any benchmark files you use, else I'll just compare against Teensy performance?

I have a few files, some of which has been added to issues, but IMO just compare against the Teensy.

> To be honest, I was hoping that you would consider upstreaming the driver into the main repos - once I get it into a state that you are happy with.

That may happen a bit later as I am considering adding you as a maintainer of the grblHAL repo. Generally speaking I do not want to add to my workload by taking over support for drivers and plugins written by others - the way grblHAL is written allows me to link to the repos containing the code instead.

**Comment by DnKon** [comment]
Hi! I cannot find H7 driver repo, are they supported?

**Reply by terjeio** [reply]
Look [here](https://github.com/dresco/STM32H7xx).

**Reply by DnKon** [reply]
Thanks!

**Comment by dresco** [comment]
Hi @terjeio

Just curious - how is your Ethernet support in this F7 driver - is it stable for you?

My H7 port is using the rewritten ST code, but am seeing an odd issue which I still need to debug. (Not receiving responses to some '?' realtime status report commands - I think it is timing sensitive, as only with some jobs, and never in check mode).

**Reply by terjeio** [reply]
I've just sent nearly 190000 blocks (10 hours) without problems via Telnet so I'll say it is stable. Check mode is slow/erratic speedwise though (random pauses).

**Reply by dresco** [reply]
Sounds a bit like what I am seeing with my websocket testing, although for me is the other way around - check mode is okay (300,000 blocks in ~20s), normal mode random pauses everywhere)..

From my initial investigations, I believe what is happening is that GRBL thinks it has sent a response, but it never appears on the wire. The session then stalls waiting for acknowledgement, until TCP timeout and re-transmission.

Adding breakpoints changes the behaviour, as does too much debug output, so think it must be timing related..

**Reply by terjeio** [reply]
Have you tried to reduce priority of the [stepper and pulse timer interrupts](https://github.com/grblHAL/STM32F7xx/blob/feeb611e2cfe68e8c00465385195732f7ef46f00/Src/driver.c#L1614-L1632)?

**Reply by dresco** [reply]
Unfortunately has made no difference.

I also included the [tcp_echoserver](https://github.com/dresco/STM32H7xx/blob/ethernet/Src/tcp_echoserver.c) sample code into my build, just for testing, and that seems to maintain a solid data transfer (up to 80Mb/s) while running the same jobs over serial.

I think I'll merge what I already have in my usb and ethernet branches into my master, and just keep digging for the root of this..

**Reply by dresco** [reply]
Am getting there, wasn't a timeout / re-transmission after all.

My issue was that link_status_callback() doesn't get called during startup - only when the cable is removed or re-inserted. So linkUp was still zero, the poll functions for the services weren't being scheduled, and the stream handler functions were only being called intermittently by the send and receive callbacks. 

Don't know if that will be the same for your F7 port - removing and re-inserting the cable should make a difference if it is?

Just looking at the logic of when the link_status and netif_status callbacks get invoked, one option might be to move the logic to the netif callback instead? But I have just run into another oddity that neither callback gets invoked when started with a static IP, so still digging...

**Reply by dresco** [reply]
Ah, not to worry. I notice you also set linkUp directly in enet_start(). Have changed my code to do something similar, and now seems to be working well for both DHCP and static IP..

**Reply by terjeio** [reply]
@dresco Does this mean your original issue went away?

**Reply by dresco** [reply]
> @dresco Does this mean your original issue went away?

Yes, was all down to the poll functions not being scheduled as expected. Appears stable now, has been running in the background for several hours with no issues.. 

I've just updated my master branch to include Ethernet support for Nucleo builds, and SD card support (via SDMMC peripheral) for the WeAct board.

**Comment by terjeio** [comment]
@dresco I have made some changes to the networking code to reduce risk for a buffer overflow and add WebDAV support:

https://github.com/grblHAL/STM32F7xx/blob/131306ee9f3decb2c9cc59986aa249c51b8cf56f/Src/enet.c#L54

https://github.com/grblHAL/STM32F7xx/blob/131306ee9f3decb2c9cc59986aa249c51b8cf56f/Src/enet.c#L146-L154

and for enable WebDAV

https://github.com/grblHAL/STM32F7xx/blob/131306ee9f3decb2c9cc59986aa249c51b8cf56f/Inc/my_machine.h#L85

There is also a new function required for WebUI support but I guess you have already implemented that:

https://github.com/grblHAL/STM32F7xx/blob/131306ee9f3decb2c9cc59986aa249c51b8cf56f/Src/enet.c#L83-L114

**Reply by dresco** [reply]
> I have made some changes to the networking code to reduce risk for a buffer overflow and add WebDAV support

Thanks, will get my H7 branch updated with latest changes..

**Reply by dresco** [reply]
fwiw - latest code gives a compilation error when building for F7 with SDcard and HTTP enabled

```
../networking/httpd.c: In function 'http_init_file':
../networking/httpd.c:1931:17: error: 'vfs_file_t' has no member named 'is_custom_file'
 1931 |         if (file->is_custom_file && (file->data == NULL))
```

**Reply by terjeio** [reply]
Oops, I forgot to commit the _.cproject_ file:

Set the` LWIP_HTTPD_CUSTOM_FILES` symbol to 0 to fix.

![image](https://user-images.githubusercontent.com/20260062/187927122-168f6fe3-ef6e-46ad-8537-556abe645d99.png)

**Reply by dresco** [reply]
Thanks..

**Reply by dresco** [reply]
Btw, I notice there were some additional LwIP memory/buffer settings added to the symbols, did they have much effect?

**Reply by terjeio** [reply]
> Btw, I notice there were some additional LwIP memory/buffer settings added to the symbols, did they have much effect?

Can't remember now - I believe I "inherited" the settings from a project I used as a template for the telnet daemon.

**Comment by dresco** [comment]
I saw the benchmarking thread on the ESP3D GitHub repo. Have just done a quick H7 test, and am seeing about 1MB/s download (is using the new ST driver) - upload is not writing any data to the card. 

SDMMC and ETH peripherals are using DMA, but they're not currently using the same memory regions - so all data is going through a scratch buffer anyway. Will try to fix that...

**Reply by terjeio** [reply]
> ... upload is not writing any data to the card

From the WebUI or from ftp and/or webdav (or all)? Any errors?

**Reply by dresco** [reply]
Hmm, looks like writes from all interfaces are failing, and possibly corrupting the card too. Think I need to revisit my SDMMC code..

**Reply by dresco** [reply]
Fix has been committed for my card corruption issues on writes (dma+cache coherency again). Have since uploaded & downloaded hundreds of MB through WebDAV, and all checksums okay. 

Performance is still poor, but have started looking into that & think there is plenty of room for improvement. Is falling back to single sector requests for a multitude of reasons - reminds me of the saying "there are a million ways for it it to go wrong, but only one way for it to go right"!

**Comment by dresco** [comment]
Hi @terjeio 

I've been doing a bit more work on networked SD card performance in my local branch, and have got to about 2MB/s for reads and 750KB/s for writes (testing with WebDAV). 

This is with a new 32GB SanDisk card, on a Nucleo dev board running at 480MHz, with 48MHz SDMMC clock. (Am also using an SD card breakout directly on the headers to avoid the need for jumper wires & slower SDMMC clock speeds).

The key seems to be trying to avoid small reads & writes via the SDMMC peripheral. These mostly happen when data is not correctly aligned for DMA access & the associated L1 cache maintenance operations.

I have had some success with this for the reads at least, but does need a patch to the networking plugin. Is this something you would consider? Will make a PR if you are agreeable, but just wanted to run the idea past you first..

Would like to do something like this;
```
diff --git a/httpd.c b/httpd.c
index 3fdfe74..0a336d1 100644
--- a/httpd.c
+++ b/httpd.c
@@ -1142,6 +1142,20 @@ static bool http_check_eof (struct altcp_pcb *pcb, struct http_state *hs)
             count = max_write_len;
 
  #endif /* HTTPD_MAX_WRITE_LEN */
+
+ #if defined(NUCLEO_H743) && L1_CACHE_ENABLE
+        /* Ensure read sizes are a multiple of 32 bytes, this helps maintain buffer alignment
+         * with L1 cache lines when performing sequential reads through a file.
+         *
+         * This results in less fallback to scratch buffer (single sector) reads, and provides
+         * a 2x read performance improvement when testing on STM32H7.
+         */
+        if (count & 0x1F) {
+            count = count & ~0x1F;
+            //printf("http_check_eof(), count adjusted to 32byte length - %i bytes\n", count);
+        }
+ #endif
+
         do {
             if ((hs->buf = (char *)mem_malloc((mem_size_t)count))) {
                 hs->buf_len = count;
```

Edit: obviously wouldn't make it Nucleo specific as above, is just my current code..

**Reply by terjeio** [reply]
A PR for this will be accepted - but make it processor specific instead of board specific?
Perhaps I should try the same for the iMXRT1062 - interesting that you gained a 2x speedup.
